### ISSUE TITLE
pxc_schedule_handler is not closing connections to ProxySQL

### ISSUE THREAD

**Issue by davidducos** [percona/pxc_scheduler_handler#47]
pxc_scheduler_handler is configure in the systemd and after start, when I run `netstat -tpn | grep pxc | wc -l` I see that the number is increasing. At the end all the memory is consumed (yes, including the swap) and ProxySQL crashes which cause to pxc_scheduler_handler to crash too. Is it possible to close that connection to avoid this issue?

Systemd unit:
```
[Unit]
Description=PXC Scheduler Handler
After=network.target proxysql.service

[Service]
Type=simple
RuntimeDirectory=proxysql
ExecStart=/apps/pxc_scheduler_handler/pxc_scheduler_handler. --configfile=config.toml --configpath=/apps/pxc_scheduler_handler/config/
PIDFile=/apps/pxc_scheduler_handler/pxc_scheduler_handler.pid
#StandardError=null  # all output is in stderr
SyslogIdentifier=pxc_sch_han
Restart=always
User=proxysql
Group=proxysql
PermissionsStartOnly=true
UMask=0007
LimitNOFILE=102400
LimitCORE=1073741824

[Install]
WantedBy=multi-user.target
```
And this configuration:
```
[pxccluster]
activeFailover = 1
failBack = false
checkTimeOut = 2000
mainSegment = 0
sslClient = "client-cert.pem"
sslKey = "client-key.pem"
sslCa = "ca.pem"
sslCertificatePath = "/opt/cert/ssl_test"
HGw = 100
HGr = 101
configHgRange =8000
maintenanceHgRange =9000
singlePrimary = true
maxNumWriters = 1
writerIsAlsoReader = 1
retryUp = 0
retryDown = 2
clusterId = 10
persistPrimarySettings=0 #0 disable| 1 only persist Write settings | 2 persist Reand and Write settings

[proxysql]
port = 6032
host = "127.0.0.1"
user="proxy_cluster_monitor"
password="PROXY_MON_PASS"
clustered = true
lockfilepath ="/apps/pxc_scheduler_handler/run"
respectManualOfflineSoft=false

[global]
debug = false
logLevel = "info"
logTarget = "file"
logFile = "/apps/pxc_scheduler_handler/pscheduler.log"
daemonize = true
daemonInterval = 2000
performance = true
OS = "na"
lockfiletimeout = 60 #seconds
lockclustertimeout = 600 #120 # seconds
```

**Comment by davidducos** [comment]
I see this ["Connection close"](https://github.com/percona/pxc_scheduler_handler/blob/main/pxc_scheduler_handler.go#L209) entry in the log.
Some netstat examples:
```
$ sudo netstat -tpn | grep pxc
tcp        0      0 127.0.0.1:60996         127.0.0.1:6032          ESTABLISHED 94729/pxc_scheduler
tcp        0      0 127.0.0.1:60954         127.0.0.1:6032          ESTABLISHED 94729/pxc_scheduler
tcp        0      0 127.0.0.1:60916         127.0.0.1:6032          ESTABLISHED 94729/pxc_scheduler
tcp        0      0 127.0.0.1:60974         127.0.0.1:6032          ESTABLISHED 94729/pxc_scheduler
tcp        0      0 127.0.0.1:60936         127.0.0.1:6032          ESTABLISHED 94729/pxc_scheduler
```

**Comment by Tusamarco** [comment]
@davidducos can you share the log?
Because i did quick check and connection is closed in https://github.com/percona/pxc_scheduler_handler/blob/0f070891ce20aba57d8edfe0fefeca5b67388c96/pxc_scheduler_handler.go#L206
so if not .. there is probably some error or exit action and need to trace where this happens

**Comment by Tusamarco** [comment]
@davidducos nevermind I reproduced it:
```
Every 1.0s: netstat -p tcp -van |grep 6032                                                                                                                                         marco-mac-work.local: Sun Jan  9 17:06:12 2022

tcp4       0      0  192.168.4.41.50529     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50536     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50541     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50545     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50549     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50553     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50557     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50565     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50570     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50574     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50582     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50586     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50590     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50594     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
tcp4       0      0  192.168.4.41.50598     192.168.4.191.6032     TIME_WAIT   131072 131768  94754      0 0x2131 0x00000008
```
but connections are in time_wait not ESTABLISHED, and closed.

**Comment by Tusamarco** [comment]
@davidducos damn i pushed comment too soon.
What I mean is that from network perspective, TCP TIME_WAIT status is just a normal behavior that after closing the session, TCP stack will hold the high port for little more time to ensure the other side receive the last FIN-ACK packet and no more data will be received in this conversation.
So by default I see the list as u mention, but it is handled and purged correctly. As such yes .. I need a log to understand what is going on.

**Comment by davidducos** [comment]
Hi @Tusamarco , I will ask the log but I saw no error, and actually I reviewed the code and reach to the point where the connection is closed. 
And yes! It is quite weird that the connection remains as ESTABLISHED. Is it possible to connect through socket instead of TCP? Just for testing...

**Comment by Tusamarco** [comment]
could be a permission issue on how the app is runned? I mean in the `Systemd unit:` definition @davidducos

**Comment by Tusamarco** [comment]
@davidducos was thinking ... 
try to run it from a real user and from shell and see if it does the same. 
To me sounds like the connection FD is not released

**Comment by davidducos** [comment]
Hi @Tusamarco, 
I run it with root and I see the same behavior, here is the output of lsof as soon as I start it:
``` 
$ sudo lsof -i | grep pxc
pxc_sched 121610          root    7u  IPv4 400590      0t0  TCP localhost:47468->localhost:6032 (ESTABLISHED)
$ sudo lsof -i | grep pxc | wc -l
3```

**Comment by davidducos** [comment]
Log: 
```
[INFO]:2022-01-10 14:01:49 - Go version: go1.16.6
[DEBUG]:2022-01-10 14:01:49 - Testing the log
[INFO]:2022-01-10 14:01:49 - Testing the log
[WARNING]:2022-01-10 14:01:49 - Testing the log
[ERROR]:2022-01-10 14:01:49 - testing log errors
[DEBUG]:2022-01-10 14:01:49 - Log initialized
[INFO]:2022-01-10 14:01:49 - LockerImpl initialized
[DEBUG]:2022-01-10 14:01:49 - ProxySQL node initialized
[DEBUG]:2022-01-10 14:01:49 - Cluster arguments  clusterid=10 hg_w:100 hg_r:101 bckhg_w:8100 bckhg_w:8101 singlePrimary:true num_writers:1 writer_is_also_r:1 retry_up:0 retry_down:2 check_timeout:2000 main_segment:0
[DEBUG]:2022-01-10 14:01:49 -  Retry up = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry down = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry up = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry down = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry up = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry down = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry up = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry down = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry up = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry down = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry up = 0
[DEBUG]:2022-01-10 14:01:49 -  Retry down = 0
[DEBUG]:2022-01-10 14:01:49 - Retrieving information from node: 192.168.23.50:3306
[DEBUG]:2022-01-10 14:01:49 - Retrieving information from node: 192.168.23.121:3306
[DEBUG]:2022-01-10 14:01:49 - Retrieving information from node: 192.168.23.49:3306
[DEBUG]:2022-01-10 14:01:49 - Retrieving information from node: 192.168.23.123:3306
[DEBUG]:2022-01-10 14:01:49 - Retrieving information from node: 192.168.23.122:3306
[DEBUG]:2022-01-10 14:01:49 - node 192.168.23.50:3306 done
[DEBUG]:2022-01-10 14:01:49 - node 192.168.23.121:3306 done
[DEBUG]:2022-01-10 14:01:49 - node 192.168.23.123:3306 done
[DEBUG]:2022-01-10 14:01:49 - node 192.168.23.49:3306 done
[DEBUG]:2022-01-10 14:01:49 - node 192.168.23.122:3306 done
[DEBUG]:2022-01-10 14:01:49 - time taken :12 checkTimeOut : 2000
[DEBUG]:2022-01-10 14:01:49 - PXC cluster data nodes initialized
[DEBUG]:2022-01-10 14:01:49 - Evaluating node 100_192.168.23.121:3306
[INFO]:2022-01-10 14:01:49 - Node: 192.168.23.121:3306 q9idajhbdb19.idm.ad.ctc HG: 100 Type W coming back ONLINE from previous OFFLINE_SOFT
[DEBUG]:2022-01-10 14:01:49 - Evaluating node 101_192.168.23.50:3306
[DEBUG]:2022-01-10 14:01:49 - Evaluating node 101_192.168.23.49:3306
[DEBUG]:2022-01-10 14:01:49 - Evaluating node 101_192.168.23.123:3306
[DEBUG]:2022-01-10 14:01:49 - Evaluating node 101_192.168.23.122:3306
[DEBUG]:2022-01-10 14:01:49 - Evaluating node 101_192.168.23.121:3306
[DEBUG]:2022-01-10 14:01:49 - Evaluating node 100_192.168.23.50:3306
[DEBUG]:2022-01-10 14:01:49 - Node 192.168.23.121:3306 is not in ONLINE state in writer HG 100 removing while evaluating
[DEBUG]:2022-01-10 14:01:49 - Evaluating for UP writer node key: 100_192.168.23.121:3306 100 192.168.23.121 3306 Status: MOVE_UP_OFFLINE
[INFO]:2022-01-10 14:01:49 - Node 100_192.168.23.121:3306 is trying to come UP in writer HG: 100 but cannot promote given we already have enough writers 1
[INFO]:2022-01-10 14:01:49 - Processing action node list and build SQL commands
[INFO]:2022-01-10 14:01:49 - Connection close
[INFO]:2022-01-10 14:01:49 - ======== Reporting execution times (nanosec/ms)by phase ============
[INFO]:2022-01-10 14:01:49 - Phase: main = 33,463,864 us 33 ms
[INFO]:2022-01-10 14:01:49 - Phase: Cluster lock = 7,607,694 us 7 ms
[INFO]:2022-01-10 14:01:49 - Phase: proxysql_init = 1,950,437 us 1 ms
[INFO]:2022-01-10 14:01:49 - Phase: main_connection = 1,031,995 us 1 ms
[INFO]:2022-01-10 14:01:49 - Phase: data_cluster_init = 22,963,596 us 22 ms
[INFO]:2022-01-10 14:01:49 - Phase: loadNodes = 9,847,345 us 9 ms
[INFO]:2022-01-10 14:01:49 - Phase: Get info for node 192.168.23.122:3306 = 12,237,126 us 12 ms
[INFO]:2022-01-10 14:01:49 - Phase: node_connection_192.168.23.122:3306 = 2,018,881 us 2 ms
[INFO]:2022-01-10 14:01:49 - Phase: Get info for node 192.168.23.50:3306 = 7,306,847 us 7 ms
[INFO]:2022-01-10 14:01:49 - Phase: node_connection_192.168.23.50:3306 = 1,399,255 us 1 ms
[INFO]:2022-01-10 14:01:49 - Phase: Get info for node 192.168.23.121:3306 = 8,396,116 us 8 ms
[INFO]:2022-01-10 14:01:49 - Phase: node_connection_192.168.23.121:3306 = 1,270,623 us 1 ms
[INFO]:2022-01-10 14:01:49 - Phase: Get info for node 192.168.23.49:3306 = 10,391,787 us 10 ms
[INFO]:2022-01-10 14:01:49 - Phase: node_connection_192.168.23.49:3306 = 1,800,516 us 1 ms
[INFO]:2022-01-10 14:01:49 - Phase: Get info for node 192.168.23.123:3306 = 8,665,304 us 8 ms
[INFO]:2022-01-10 14:01:49 - Phase: node_connection_192.168.23.123:3306 = 1,729,909 us 1 ms
[INFO]:2022-01-10 14:01:49 - Phase: Get Action Map (DataClusterImpl) = 755,030 us 0 ms
[INFO]:2022-01-10 14:01:49 - Phase: Process changes - ActionMap - (ProxysqlNode) = 8,231 us 0 ms
[INFO]:2022-01-10 14:01:51
```

As you can see, there is a `Connection close` there but for some reason, it is not being closed...

**Comment by davidducos** [comment]
@Tusamarco, 
I think that I found the root cause, but I need you to confirm. 
We have a proxysql cluster! but if I disable it, I don't see the connection established anymore. So, what I thinking that it is happening, is that you are closing just the first connection in the list, and not the connections opened to all the proxysql servers. Is that possible? should I be getting 2 `Connection close` in the log?

**Comment by impimp** [comment]
It comes from:

https://github.com/percona/pxc_scheduler_handler/blob/main/internal/DataObjects/Locker.go#L118-L119

that's creating the new connection each iteration. The only workaround is to have clustered=false.

**Comment by Tusamarco** [comment]
yes @impimp already found the bug and fixing it

**Comment by Tusamarco** [comment]
@davidducos I have created a fix in this branch https://github.com/percona/pxc_scheduler_handler/tree/Issue-47
I am now running the unit tests and MTR test to prevent regressions but if u can test it as well I will be more than happy.

From my testing I can see now all connection in TIME_WAIT.
Let me know

**Comment by Tusamarco** [comment]
@davidducos  fixed in [v1.3.2](https://github.com/percona/pxc_scheduler_handler/releases/tag/v1.3.2)
Confirm and please close the issue

**Comment by Tusamarco** [comment]
Fixed closing
